### PR TITLE
cmd: add `snap debug validate-seed <path>` cmd

### PR DESCRIPTION
**Pull request by mvo5**
We had a bug where one of the recent livecd-rootfs build images
had an incorrect seed.yaml. This caused a panic of snapd on when
seeding. To prevent this here is a small helper that we can run
inside livecd-rootfs to ensure seed.yaml is good enough to not
crash snapd.

This is for https://bugs.launchpad.net/ubuntu/+source/livecd-rootfs/+bug/1825437

### CODE REVIEW

File: cmd/snap/cmd_debug_validate_seed.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+type cmdValidateSeed struct {
+	Positionals struct {
+		SeedYamlPath string `positional-arg-name:"<seed-yaml-path>"`
+	} `positional-args:"true"`
+}
+
+func init() {
+	cmd := addDebugCommand("validate-seed",
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_validate_seed.go]: or `validate-seed-file` ?

File: cmd/snap/cmd_debug_validate_seed.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		}, nil, nil)
+	cmd.hidden = true
+}
+
+func (x *cmdValidateSeed) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	if _, err := snap.ReadSeedYaml(x.Positionals.SeedYamlPath); err != nil {
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_validate_seed.go]: When seeing, a bunch of checks is done after we load the seed.yaml file. Should we validate at least the name and channel here as well?

File: cmd/snap/cmd_debug_validate_seed_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	err := ioutil.WriteFile(tmpf, []byte(`
+snaps:
+ -
+   name: core
+   channel: foo
+   file: bar
+ -
+   name: gnome-foo
+   channel: foo
+   name: xxx
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_validate_seed_test.go]: duplicate `name`

File: cmd/snap/cmd_debug_validate_seed_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+snaps:
+ -
+   name: core
+   channel: foo
+   file: bar
+ -
+ -
+   name: gnome-foo
+   channel: foo
+   name: xxx
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_validate_seed_test.go]: same here

File: cmd/snap/cmd_debug_validate_seed.go (hunk 2 above)
Comment (stolowski) [inline cmd/snap/cmd_debug_validate_seed.go]: Not super important, but would it make sense to make seed yaml path argument optional, and use the default /var/lib/snapd/.../seed.yaml path if not provided, for convienience?

Comment (mvo5) [reply cmd/snap/cmd_debug_validate_seed.go]: Good point, thanks! I updated to the PR add basic checks.

File: cmd/snap/cmd_debug_validate_seed_test.go (hunk 3 above)
Comment (mvo5) [reply cmd/snap/cmd_debug_validate_seed_test.go]: Thanks, fix this (and the other one) with more realistic inputs.

File: cmd/snap/cmd_debug_validate_seed.go (hunk 1 above)
Comment (mvo5) [reply cmd/snap/cmd_debug_validate_seed.go]: Not sure, feel a bit long but I don't really mind strongly either way. Its a debug command after all :)

File: cmd/snap/cmd_debug_validate_seed.go (hunk 2 above)
Comment (mvo5) [reply cmd/snap/cmd_debug_validate_seed.go]: I like the idea, but I'm also a bit lazy :) We could do it in a followup.

Comment (stolowski) [reply cmd/snap/cmd_debug_validate_seed.go]: Sure, no problem.

File: snap/seed_yaml.go [hunk 5]
```diff
@@ -58,20 +58,36 @@ type Seed struct {
 }
 
 func ReadSeedYaml(fn string) (*Seed, error) {
+	errPrefix := "cannot read seed yaml: "
+
 	yamlData, err := ioutil.ReadFile(fn)
 	if err != nil {
-		return nil, fmt.Errorf("cannot read seed yaml: %s", fn)
+		return nil, fmt.Errorf(errPrefix+"%v", err)
```
Comment (stolowski) [inline snap/seed_yaml.go]: Nitpick: looks a bit weird to concatenate errorPrefix that way when using fmt formatting anyway, why not simply `fmt.Errorf("%s%v", errorPrefix, err)` (also below)?

File: snap/seed_yaml.go [hunk 6]
```diff
@@ -58,20 +58,36 @@ type Seed struct {
...
 
 	// validate
 	for _, sn := range seed.Snaps {
+		if sn == nil {
+			return nil, fmt.Errorf(errPrefix + "empty element in seed")
+		}
+		if err := ValidateInstanceName(sn.Name); err != nil {
+			return nil, fmt.Errorf(errPrefix+"%v", err)
+		}
+		if sn.Channel != "" {
```
Comment (stolowski) [inline snap/seed_yaml.go]: Nice

**Comment by bboozzoo** [comment]
Shall we land it?

**Pull request merged**